doc/man3/COMP_CTX_new.pod

```diff
@@ -0,0 +1,175 @@
+buffer I<out> of size I<olen> using the lgorithm specified by I<ctx>.
+
+Methods (B<COMP_METHOD>) may be specified by one of these functions. These functions
+will be available even if their corresponding compression algorithm is not configured
+into the OpenSSL library. In such a case, a non-operative method will be returned.
+
+=over 4
+
+=item *
+
+COMP_zlib() returns a B<COMP_METHOD> for stream-based ZLIB compression.
+
+=item *
+
+COMP_brotli() returns a B<COMP_METHOD> for stream-based Brotli compression.
+
+=item *
+
+COMP_brotli_oneshot() returns a B<COMP_METHOD> for one-shot Brotli compression.
+
+=item *
+
+COMP_zstd() returns a B<COMP_METHOD> for stream-based Zstandard compression.
+
+=item *
+
+COMP_zstd_oneshot() returns a B<COMP_METHOD> for one-shot Zstandard compression.
+
+=back
+
+BIO_f_zlib(), BIO_f_brotli() BIO_f_zstd() each return a B<BIO_METHOD> that may be used to
+create a B<BIO> via B<BIO_new(3)> to read and write compressed files or streams.
+The functions are only available if the corresponding algorithm is compiled into
+the OpenSSL library. NULL may be returned if the algorithm fails to load dynamically.
+
+=head1 NOTES
+
+While compressing non-compressible data, the output may be larger than the
+input. Care should be taken to size output buffers appropriate for both
+compression and expansion.
+
+Compression support and compression algorithms must be enabled and built into
+the library before use. Refer to the INSTALL file when configuring OpenSSL.
+
+ZLIB may be found at L<https://zlib.net>
+
+Brotli may be found at L<https://github.com/google/brotli>.
+
+Zstandard may be found at L<https://github.com/facebook/zstd>.
+
+Compression of SSL/TLS records is not recommended, as it has been
+shown to lead to the CRIME attack L<https://en.wikipedia.org/wiki/CRIME>.
+It is disabled by default, and may be enabled by clearing the
+SSL_OP_NO_COMPRESSION options of the L<SSL_CTX_set_options(3)> or
+L<SSL_set_options(3)> functions.
+
+Compression is also used to support certificate compression as described
+in RFC8879 L<https://datatracker.ietf.org/doc/html/rfc8879>.
+It may be disabled via the SSL_OP_NO_TX_CERTIFICATE_COMPRESSION and
+SSL_OP_NO_RX_CERTIFICATE_COMPRESSION options of the
+L<SSL_CTX_set_options(3)> or L<SSL_set_options(3)> functions.
+
+COMP_zlib(), COMP_brotli() and COMP_zstd() are stream-based compression methods.
+Internal state (including compression dictionary) is maintained between calls.
+If an error is returned, the stream is corrupted, and should be closed.
+
+COMP_brotli_oneshot() and COMP_zstd_oneshot() are not stream-based. These
+methods do not maintain state between calls. An error in one call does not affect
+future calls.
+
+=head1 RETURN VALUES
+
+COMP_CTX_new() returns a B<COMP_CTX> on success, or NULL on failure.
+
+COMP_CTX_get_method(), COMP_zlib(), COMP_brotli(), COMP_brotli_oneshot(),
+COMP_zstd(), and COMP_zstd_oneshot() return a B<COMP_METHOD> on success,
+or NULL on failure.
```

Comment:
"Right" is questionable. ZLIB always returned *something; a non-operative `COMP_METHOD` was returned if ZLIB was not enabled.
If you are thinking it ought to return NULL, I'm in agreement with you, but I was following existing code behavior. If the thought is that it should return NULL, rather than a non-operative `COMP_METHOD`, I'd support doing that in this PR.

Reply:
This comment is still outstanding. Docs need updating to say they return an "empty" method on failure - or something like that

Reply:
The code was changed to return NULL on error for `COMP_zlib()`and friends. All functions that accept `COMP_METHOD` check for NULL, and being an opaque structure, users could never dereference `COMP_METHOD`. So, thought ought to be safe. The docs state that NULL is returned on failure. Not sure what else you want here?